File: demo/src/main/java/com/novoda/merlin/demo/presentation/DemoActivity.java
package com.novoda.merlin.demo.presentation;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.novoda.merlin.Bindable;
import com.novoda.merlin.Connectable;
import com.novoda.merlin.Disconnectable;
import com.novoda.merlin.Merlin;
import com.novoda.merlin.MerlinsBeard;
import com.novoda.merlin.NetworkStatus;
import com.novoda.merlin.demo.R;
import com.novoda.merlin.demo.connectivity.display.NetworkStatusDisplayer;
import com.novoda.merlin.demo.presentation.base.MerlinActivity;

public class DemoActivity extends MerlinActivity implements Connectable, Disconnectable, Bindable {

    private NetworkStatusDisplayer networkStatusDisplayer;
    private MerlinsBeard merlinsBeard;
    private View viewToAttachDisplayerTo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);

        viewToAttachDisplayerTo = findViewById(R.id.displayerAttachableView);
        merlinsBeard = new MerlinsBeard.Builder()
                .build(this);
        networkStatusDisplayer = new NetworkStatusDisplayer(getResources(), merlinsBeard);

        findViewById(R.id.current_status).setOnClickListener(networkStatusOnClick);
        findViewById(R.id.has_internet_access).setOnClickListener(hasInternetAccessClick);
        findViewById(R.id.wifi_connected).setOnClickListener(wifiConnectedOnClick);
        findViewById(R.id.mobile_connected).setOnClickListener(mobileConnectedOnClick);
        findViewById(R.id.network_subtype).setOnClickListener(networkSubtypeOnClick);
        findViewById(R.id.next_activity).setOnClickListener(nextActivityOnClick);
    }

    private final View.OnClickListener networkStatusOnClick = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            if (merlinsBeard.isConnected()) {
                networkStatusDisplayer.displayPositiveMessage(R.string.current_status_network_connected, viewToAttachDisplayerTo);
            } else {
                networkStatusDisplayer.displayNegativeMessage(R.string.current_status_network_disconnected, viewToAttachDisplayerTo);
            }
        }
    };

    private final View.OnClickListener hasInternetAccessClick = new View.OnClickListener() {
        @Override
        public void onClick(final View view) {
            merlinsBeard.hasInternetAccess(new MerlinsBeard.InternetAccessCallback() {
                @Override
                public void onResult(boolean hasAccess) {
                    if (hasAccess) {
                        networkStatusDisplayer.displayPositiveMessage(R.string.has_internet_access_true, viewToAttachDisplayerTo);
                    } else {
                        networkStatusDisplayer.displayNegativeMessage(R.string.has_internet_access_false, viewToAttachDisplayerTo);
                    }
                }
            });
        }
    };

    private final View.OnClickListener wifiConnectedOnClick = new View.OnClickListener() {

        @Override
        public void onClick(View view) {
            if (merlinsBeard.isConnectedToWifi()) {
                networkStatusDisplayer.displayPositiveMessage(R.string.wifi_connected, viewToAttachDisplayerTo);
            } else {
                networkStatusDisplayer.displayNegativeMessage(R.string.wifi_disconnected, viewToAttachDisplayerTo);
            }
        }
    };

    private final View.OnClickListener mobileConnectedOnClick = new View.OnClickListener() {

        @Override
        public void onClick(View view) {
            if (merlinsBeard.isConnectedToMobileNetwork()) {
                networkStatusDisplayer.displayPositiveMessage(R.string.mobile_connected, viewToAttachDisplayerTo);
            } else {
                networkStatusDisplayer.displayNegativeMessage(R.string.mobile_disconnected, viewToAttachDisplayerTo);
            }
        }
    };

    private final View.OnClickListener networkSubtypeOnClick = new View.OnClickListener() {

        @Override
        public void onClick(View view) {
            networkStatusDisplayer.displayNetworkSubtype(viewToAttachDisplayerTo);
        }
    };

    private final View.OnClickListener nextActivityOnClick = new View.OnClickListener() {

        @Override
        public void onClick(View view) {
            Intent intent = new Intent(getApplicationContext(), DemoActivity.class);
            startActivity(intent);
        }
    };

    @Override
    protected Merlin createMerlin() {
        return new Merlin.Builder()
                .withConnectableCallbacks()
                .withDisconnectableCallbacks()
                .withBindableCallbacks()
                .build(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        registerConnectable(this);
        registerDisconnectable(this);
        registerBindable(this);
    }

    @Override
    public void onBind(NetworkStatus networkStatus) {
        if (!networkStatus.isAvailable()) {
            onDisconnect();
        }
    }

    @Override
    public void onConnect() {
        networkStatusDisplayer.displayPositiveMessage(R.string.connected, viewToAttachDisplayerTo);
    }

    @Override
    public void onDisconnect() {
        networkStatusDisplayer.displayNegativeMessage(R.string.disconnected, viewToAttachDisplayerTo);
    }

    @Override
    protected void onPause() {
        super.onPause();
        networkStatusDisplayer.reset();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        networkStatusDisplayer = null;
    }
}

File: demo/src/main/java/com/novoda/merlin/demo/presentation/DemoActivity.kt
package com.novoda.merlin.demo.presentation

import android.os.Bundle
import android.view.View
import com.novoda.merlin.Merlin
import com.novoda.merlin.MerlinsBeard
import com.novoda.merlin.NetworkStatus
import com.novoda.merlin.demo.R
import com.novoda.merlin.demo.connectivity.display.NetworkStatusDisplayer
import com.novoda.merlin.demo.presentation.DemoActivity.DisplayStrategy.*
import com.novoda.merlin.demo.presentation.base.MerlinActivity
import com.novoda.merlin.registerable.bind.Bindable
import com.novoda.merlin.registerable.connection.Connectable
import com.novoda.merlin.registerable.disconnection.Disconnectable
import kotlinx.android.synthetic.main.main.displayerAttachableView
import kotlinx.android.synthetic.main.main.current_status as currentStatus
import kotlinx.android.synthetic.main.main.mobile_connected as mobileConnected
import kotlinx.android.synthetic.main.main.network_subtype as networkSubtype
import kotlinx.android.synthetic.main.main.wifi_connected as wifiConnected

class DemoActivity : MerlinActivity(), Connectable, Disconnectable, Bindable {

    override val merlin: Merlin by lazy {
        Merlin.Builder()
            .withConnectableCallbacks()
            .withDisconnectableCallbacks()
            .withBindableCallbacks()
            .build(this)
    }

    private val merlinsBeard by lazy {
        MerlinsBeard.from(this)
    }

    private val networkStatusDisplayer by lazy {
        NetworkStatusDisplayer(resources, merlinsBeard)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)

        currentStatus.onClickDisplay(IfConnected)
        wifiConnected.onClickDisplay(IfWifiConnected)
        mobileConnected.onClickDisplay(IfMobileConnected)
        networkSubtype.onclickDisplaySubtype()
    }

    private fun View.onclickDisplaySubtype() {
        setOnClickListener {
            networkStatusDisplayer.displayNetworkSubtype(this)
        }
    }

    private fun View.onClickDisplay(displayStrategy: DisplayStrategy) {
        setOnClickListener {
            networkStatusDisplayer.displayNegativeMessage(
                displayStrategy.message,
                displayerAttachableView
            )
        }
    }

    private val DisplayStrategy.message get() =
        if (merlinsBeard.isPositive()) positiveMessage else negativeMessage

    private sealed class DisplayStrategy(
        val positiveMessage: Int,
        val negativeMessage: Int,
        val isPositive: MerlinsBeard.() -> Boolean
    ) {

        object IfConnected : DisplayStrategy(
            R.string.current_status_network_connected,
            R.string.current_status_network_disconnected,
            MerlinsBeard::isConnected
        )

        object IfWifiConnected : DisplayStrategy(
            R.string.wifi_connected,
            R.string.wifi_disconnected,
            MerlinsBeard::isConnectedToWifi
        )

        object IfMobileConnected : DisplayStrategy(
            R.string.mobile_connected,
            R.string.mobile_disconnected,
            MerlinsBeard::isConnectedToMobileNetwork
        )

    }

    override fun onResume() {
        super.onResume()
        registerConnectable(this)
        registerDisconnectable(this)
        registerBindable(this)
    }

    override fun onConnect() {
        networkStatusDisplayer.displayPositiveMessage(R.string.connected, displayerAttachableView)
    }

    override fun onDisconnect() {
        networkStatusDisplayer.displayNegativeMessage(R.string.disconnected, displayerAttachableView)
    }

    override fun onBind(networkStatus: NetworkStatus) {
        if (!networkStatus.isAvailable) {
            onDisconnect()
        }
    }

    override fun onPause() {
        super.onPause()
        networkStatusDisplayer.reset()
    }

}
